Feature: implement more intuitive y-intervals using subdivide

hhpmmd opened the pull request:
Regards #68 

Here's my (first) approach to solving this. It basically checks if using fewer or more subdividing y-intervals lead to integer (for numerical data) or 5-min (for time data) intervals (after typing this i realize that for y-ranges below 30 mins, some work could also be done). 

It uses a Pair which might be bad practice, but making an extra class for it seemed a bit extra.

One thing I noticed it that it seems kind of random how many intervals the algorithm will choose (for the end user). Maybe that should be improved, not sure though.

hhpmmd commented:
Here are some screenshots:


![zz0](https://user-images.githubusercontent.com/77058867/106817451-2bfbdd80-6677-11eb-86a2-64f6945a8563.png)
![zz1](https://user-images.githubusercontent.com/77058867/106817456-2dc5a100-6677-11eb-98d3-b38838f5b814.png)
![zz2](https://user-images.githubusercontent.com/77058867/106817466-2ef6ce00-6677-11eb-9172-b2fe4b35d157.png)
![zz3](https://user-images.githubusercontent.com/77058867/106817474-30c09180-6677-11eb-8978-522d64d3539f.png)
![zz4](https://user-images.githubusercontent.com/77058867/106817476-31f1be80-6677-11eb-8976-b641f125fd5d.png)

hhpmmd commented:
Note: The app crashes when there's an error during `val y_range = bounds.height.toDouble()`. One case where this happens is when there's not enough data to plot. I catch this case using a check whether `bounds.minY == null`. Maybe it makes sense to use a catch statement here, to be sure in case there might be other ways the `height.toDouble()` call can fail.

SamAmco commented:
This looks like a really good start. I want to take a little time to review and test my end so I may get back to you further on the weekend.

hhpmmd commented:
This approach works well in fixed bounds mode, but struggles in the dynamic mode.

One idea to have the benefits of dynamic mode (not having to readjust the bounds when new min/max values are added to the data) and of fixed mode (nice y-intervals) would be to maybe have a third bounds mode _dynamic (with margins)_ (open for other names) which is dynamic in the sense that the top and bottom bounds are automatically updated when there are new min/max values in the data, but some margins are added so that the top and bottom bounds are not exactly the min/max data but rather mutiples of 1 or 10 if the max is larger, so that the new subdivide-inverval-algorithm has an easy time subdividing the y-range.

So if we have data with min=2.7 and max=7.6

- with dynamic bounds: algorithm won't be able to do anything
- with dynamic (with margins) bounds: min_bound = 2, max_bound=8 -> algorithm finds nice intervals

SamAmco commented:
I think it's quite a tough problem to solve completely so I'm open to partial solutions and minor improvements. It may be that what you have so far is enough to start with. You are right though that there are a couple of gaps in it:

1. As you say the data range must be exact. I would like to avoid adding more range options if possible. Especially as this seems like something that should definitely be possible to do automatically. I wrote an implementation that allows for some un-used range:

```    private fun getYAxisParameters(
        bounds: RectRegion,
        durationBasedRange: Boolean
    ): Pair<StepMode, Double> {
        val targetIntervals = 11.0

        if (bounds.minY == null) // when there's not enough data to plot
            return Pair(StepMode.SUBDIVIDE, targetIntervals)

        val y_range = bounds.height.toDouble()

        val moduloVal = if (durationBasedRange) 60 * 5 else 1

        //We can accept a certain amount of the range not being used if it allows us to have a neat
        // number of divisions in the range. For example if the data is between 0 and 6.97 then we
        // might as well have a Y axis with the numbers 0 to 7
        val acceptableUnusedRangePercentage = 0.1
        fun Double.closeToNeatDivisor() =
            this == 0.0 || (this < 1.0 && this > (1.0 - acceptableUnusedRangePercentage))

        //We don't want to have too many or too few divisions on the axis so we only give a custom
        //number of divisions if it is close to the target amount
        val max_divergence = 4
        for (d in 0..max_divergence) {
            if ((y_range / (targetIntervals - 1 - d)).rem(moduloVal).closeToNeatDivisor())
                return Pair(StepMode.SUBDIVIDE, targetIntervals - d)
            if ((y_range / (targetIntervals - 1 + d)).rem(moduloVal).closeToNeatDivisor())
                return Pair(StepMode.SUBDIVIDE, targetIntervals + d)
        }
        return Pair(StepMode.SUBDIVIDE, targetIntervals)
    }
```

However this is not a complete fix because it modifies the number of range divisions without actually modifying the range of the data So if you create a data set with two points at 1 and 6.9, you get 7 divisions (great) but the graph still shows 6.9 at the top like so: 

![Screenshot_20210207_174826_com samco trackandgraph debug](https://user-images.githubusercontent.com/5012146/107154844-fbfd5480-696c-11eb-9054-e3796bb2a55a.jpg)

The code must of course make sure that it only ever adds range in order to get neat divisions and never subtracts range. 

Anyway maybe you can expand on this to fix the issue of the data bounds not being modified. 

2. This one is less of an issue but suppose I have data in the range 0-70 then your implementation would give 10 divisions rather than 7 because it searches outwards from the "ideal" of 10. Personally I think having 7 divisions with increments of 10 is nicer in this scenario. Perhaps the general rule should be that the smallest number of divisions (within the maximum range) is the ideal? I'm not certain of this though, what do you think? It seems like you could easily change this by iterating with something like:

```        
       for (d in -max_divergence..max_divergence) {
            if ((y_range / (targetIntervals - 1 + d)).rem(moduloVal).closeToNeatDivisor())
                return Pair(StepMode.SUBDIVIDE, targetIntervals + d)
        }
```

3. None of these implementations will work for ranges smaller than 1 right now. So for example I like to track my weight which thankfully does not vary much more than 1kg from month to month so the range might show values between say 75 and 75.7 .. Here of course it would be nice to have 7 divisions but I suspect we will get 10. Maybe this can be fixed by multiplying the y range by 10 until you get a whole number and using that to calculate the number of desired divisions. 


Let me know if you have any other thoughts. If you don't have time to work on it further I will probably merge what you've already done anyway because it's still an improvement and I don't think it breaks anything. Thanks :)

hhpmmd commented:
Regarding 1: I'll have a look at it.

Regarding 2 & 3: 

I think they have the same underlying problem. There is a preference to have steps of size 1, 2, 2.5, 5 * 10*z as steps and not just an integer number (what is currently implemented). A similar thing is true for time data. 

I think it would be a good approach to check if a division that results in such intervals is possible (as you describe in 2). It becomes slightly more complex when we can also modify the bounds as the number of possible bounds/interval combinations gets bigger. 

I think the code will probably get pretty a bit messier and a little brute-force-ish, but it should remain within a reasonable degree.

I'll probably have something before the end of the week.

hhpmmd commented:
Okay, I think I came up with a good general algorithm for this. It's currently in python since that's what I prototyped with, but I'll port it to kotlin over the weekend.

Here's it's result for the two scenarios you describe earlier: _(read [min -> max] number_intervals × interval_size, and then the y-intervals (from bot to top))_

```
[0.0 -> 70.0] 7×12.5
-- 0.0
-- 12.5
-- 25.0
-- 37.5
-- 50.0
-- 62.5
-- 75.0
```
Here it selects the the 12.5 interval instead of the 10 interval because it has a bias to select the larger interval to get a smaller number of total intervals (choice here between 7 × 12.5 or 8 × 10). I have an idea how it could be adapted and I'll look at it, but in the meantime this should also be acceptable.
```
[75.0 -> 75.7] 8×0.1
-- 75.0
-- 75.1
-- 75.2
-- 75.3
-- 75.4
-- 75.5
-- 75.6
-- 75.7
```
Here's the algorithm in python and some more examples in a txt file

[track_and_graph_yintervals.py.txt](https://github.com/SamAmco/track-and-graph/files/5974446/track_and_graph_yintervals.py.txt)
[track_and_graph_yintervals_examples.txt](https://github.com/SamAmco/track-and-graph/files/5974447/track_and_graph_yintervals_examples.txt)

(Had to rename the py file to upload it here..)

If you find any example in the file where you think a different interval should be better, please point it out, or if you have more general feedback about the solution.

hhpmmd commented:
So this would be a new dynamic range style, it shouldn't be to hard to adapt it to the fixed variant.

SamAmco commented:
The results look good. I'm not sure I fully followed your code. It would be good to have some comments to guide me through when you port it. Also I think it would be a good idea to put this function somewhere general (like Statistics.kt for now although I need to refactor this really). I think it's likely that at some point in the future there may be new graph types like bar charts or stacked graphs that also have Y-axis' and need the same code. This is probably also a good example of code that could/should be unit tested.

My only question mark on the outputs really was that I'm not sure what the limit of un-used Y-range actually is. So in my example code above I suggested that if more than 10% of the range is not used then we should find a new divisor. In your second example: 

```
[11:14:45 -> 15:07:38] 6×1:00:00
-- 11:00:00
-- 12:00:00
-- 13:00:00
-- 14:00:00
-- 15:00:00
-- 16:00:00
```

There are 5 hours of Y range but the actual range of the data is 3:52:53 so about 78% of the range is used. Using 78% of the range is not necessarily a problem and no doubt a small price to pay for a consistent algorithm, but I just want to make sure you have considered the extreme cases? i.e. what is the maximum amount of un-used range this method could produce?

hhpmmd commented:
Okay, I finally found some time to port the idea to kotlin an embed it in the app. I have briefly tested it using the emulator and it seems to work.

You made a good point about the unused range so I adapted the algorithm to keep track of that. There is a test that runs a bunch of random numbers through the algorithm and these are the % of ranges used:

```
                1.0  -> 0.97:  4.690510846745976%
                0.97 -> 0.94: 13.121357592722184%
                0.94 -> 0.91: 20.35140657802659%
                0.91 -> 0.88: 26.00474457662701%
                0.88 -> 0.85: 20.37006298110567%
                0.85 -> 0.82:  9.145878236529041%
                0.82 -> 0.79:  4.100265920223933%
                0.79 -> 0.76:  1.569447165850245%
                0.76 -> 0.73:  0.527851644506648%
                0.73 -> 0.7:   0.06376487053883835%
```
meaning in 4.7% of all cases it uses between 100% and 97% of the range and so on. 

It's possible to adapt some hyper-parameters to shift these values around or we could just use the old way if the range used is below 79% or something. I'll check what these lower cases look like in the app to get a feeling for what to do.

hhpmmd commented:
Here are some more examples 
[new_examples_numerical.txt](https://github.com/SamAmco/track-and-graph/files/6002715/new_examples_numerical.txt)
[new_examples_time.txt](https://github.com/SamAmco/track-and-graph/files/6002716/new_examples_time.txt)

hhpmmd commented:
Let me know if the code is easier to understand, I guess the helper functions inside the main function make it kind of messy, I'll probably move them out / rename them before the merge.

You should also check if there should be things like `internal` or `suspended` before the functions; I didn't really understand what they do / where I should use them.

hhpmmd commented:
I cleaned up the code and added some more evaluation.

I also took a closer look at the the cases in which they we get poor y-range usage. Like this one:
```
110.0
105.0
100.0
95.0
90.0
85.0
80.0
75.0
70.0
65.0
Data range: [69.3 -> 105.2] interval: 5.0 x 9 | range used = 79.8
```
Pretty much all of these 'bad' cases have in common that the intervals at the top and bottom are barely full. Here we use only 14% of the 65-70 range and 4% of the 105-110 range. One idea would be to drop these intervals at the borders if they contain less than a certain amount if data. For example if we were to drop intervals which are filled less than 5% the new result would be:
```
           <- this interval disappeared
105.0
100.0
95.0
90.0
85.0
80.0
75.0
70.0
65.0
Data range: [69.3 -> 105.2] interval: 5.0 x 8 | range used = 89.8 <- a 10% increase
```

I think it's a risky topic, but I also think a small threshold should be reasonable. You could even go so far to say that the latest datapoint should never be pushed out of the plot this way, but it's not so important for older data. What are your thoughts on this?

hhpmmd commented:
Also some of the tests use random sampling of start/end data, making them non-deterministic. It shouldn't be a problem here, but might be bad code style.

SamAmco commented on app/src/test/java/com/samco/trackandgraph/statistics/Statistics_getYParameters_KtTest.kt:
```diff
@@ -0,0 +1,312 @@
+            true  -> (1..72000).map { it.toDouble() } // one second to 20 hours
+        }
+
+        for (start in startValues.shuffled().slice(0..10000)) {
```
I appreciate knowing that the code is this consistent but I think this test is a little annoyingly slow to run in quick succession. I suggest we reduce iterations on these tests considerably.

hhpmmd replied:
Yeah, maybe add a flag whether you want to do an excessive test or just a quick check, with the quick one being the default.

SamAmco commented on app/src/main/java/com/samco/trackandgraph/statistics/Statistics.kt:
```diff
@@ -557,3 +558,139 @@ internal fun getLargestBin(bins: List<List<Double>>?): Double? {
+data class YAxisParameters (    val step_mode: StepMode, val n_intervals: Double,
+                                val bounds_min: Double, val bounds_max: Double,)
+
+fun finishProto (proto: PossibleIntervalProto, y_min: Double, y_max: Double, fixedBounds: Boolean): PossibleInterval {
```
Some of these functions look like they could be private to me. I would appreciate if you could change this because until this file is refactored it helps to know which functions are actually called by other classes.

SamAmco reviewed (changes requested):
![Screenshot_20210222_183259_com samco trackandgraph debug](https://user-images.githubusercontent.com/5012146/108753443-a7e19b00-753c-11eb-9a67-f9aa4a0ed943.jpg)

The only issue I noticed was that i created a data set with only 2 points with values 0.01 and 0.09 and then created a graph of those (above). I'm not entirely sure where the flaw lies here. There is a related issue which is that the precision of the Y labels seems to always be 1 decimal place right now which doesn't help when identifying the issue. Perhaps this output makes more sense to you?

I put a few comments in my review but they're all small things. Overall this code is looking very good. There is no doubt some work to be done to refactor Statistics.kt since none of that code really belongs together anymore but I can do that at a later time.

hhpmmd commented:
I did the same plot you did in the current version (the one without my changes) and it produces the following result (**with dynamic setting!**):
![image](https://user-images.githubusercontent.com/77058867/108757162-c601c800-7549-11eb-85e0-0245a5b5aa8e.png)

So there has to be some complication with numbers this small from the plotting side. It makes sense that the y-labels are rounded the way they are, but why is the maximum value not at the top of the plot, like it always is with the old dynamic behavior? :shrug: 

On a different note, I thought about changing the y-label formatting if the number is an integer to drop the decimal place, not sure how easy that is to do.

hhpmmd commented on app/src/main/java/com/samco/trackandgraph/graphstatview/decorators/GraphStatLineGraphDecorator.kt:
```diff
@@ -111,9 +111,9 @@ class GraphStatLineGraphDecorator(listMode: Boolean) :
 
     private suspend fun setLineGraphBounds() {
         val bounds = data!!.bounds
-        if (data!!.yRangeType == YRangeType.FIXED) {
```
I wasn't aware of this side effect. Making it dependent on the listView sounds good to me.

hhpmmd replied:
**So do you also want to not hard-set bounds in fullscreen-mode when using the fixed mode?** I feel that if you choose to view it fullscreen, you might want to scroll and not be hindered by the bounds, even though you set them (by using fixed mode).

SamAmco replied:
It's kind of tricky actually because sometimes it's more irritating that the bounds move because the lines jump around and it can be confusing. I find It depends on the nature of the data you're looking at. I think ideally you should be able to turn this on and off in the full screen view but for now it may be best to keep the behaviour as is and have it only work if the bounds are dynamic mode.

hhpmmd commented on app/src/main/java/com/samco/trackandgraph/graphstatview/factories/LineGraphDataFactory.kt:
```diff
@@ -163,6 +169,34 @@ class LineGraphDataFactory : ViewDataFactory<LineGraphWithFeatures, ILineGraphVi
+    }
+
+
     private suspend fun getBounds(
```
There is this yield call I didn't copy into the new function, not sure what exactly it does and if it has a place in the new code. Besides that, it can be removed.

SamAmco replied:
yield and suspend are both related to the use of kotlin co-routines ( https://kotlinlang.org/docs/coroutines-overview.html ) .. I don't think you need the yield here really but I may add it if I notice a slow down. It's just about breaking the work up into chunks so as not to steal too many resources at once and cause frame drops.

hhpmmd commented on app/src/main/java/com/samco/trackandgraph/statistics/Statistics.kt:
```diff
@@ -557,3 +558,139 @@ internal fun getLargestBin(bins: List<List<Double>>?): Double? {
+                   fixedBounds: Boolean,
+                   throw_exc_if_non_found: Boolean = false) : YAxisParameters {
+
+    fun convertToOutput(interval: PossibleInterval) = YAxisParameters(
```
The way I used it before it made sense, and I didn't disliked it enough to change. I'll inline it.

SamAmco commented:
Yes well spotted. I should have checked that my self really, sorry. Probably best to leave the range precision issue out of this PR and fix it in a new one (if you're interested) .. I think it's probably a lot simpler than the problem you've already solved. Just a question of looking at the range and then deciding how many decimal places is most appropriate.

hhpmmd commented:
Okay so I wanted to see what a lot of plots look like in the app and wasn't sure what an easy way to do this was. I found an easy way (and I think it works great for what I wanted to see), but I'm not sure if the changes should stay in the codebase.

So what I did is I changed to behavior of the duplicate button to not produce a 1 to 1 copy, but rather a 1 to 33 copy where the copies also have random scale and offset, leading to a wide range of plots where you can actually see the new y-intervals in action. Right now the numbers (offset and scale) are finetuned for numerical data; if you want to do the same for time data you have to adjust the function.

Again, I think it is very helpful in the current situation, but also very hacky and should be reverted as before the merge. Maybe in the long run it makes sense to have extra options for special duplications in the debug app, however that looks like it needs a fair bit of refactoring / changing the code (as far as I could tell).

SamAmco commented:
That's an interesting idea. I think yes this code should not be in master. Ideally I would not have any un-used/commented code in the repo. Especially if it is old code I think it is best to remove it all together since it is still in the git history anyway. In this case I would say the simplest thing for you to do in the short term may be just to create a local branch and add your debug code in that. As you modify your feature branch you can occasionally switch back to your debug branch and merge the latest from your feature branch to test it with the debug code. 

Longer term i'm not sure what's best. As the app grows in complexity, the lack of proper integration tests is of growing concern and the QA process is not currently very good. I think it definitely makes sense to have some code that generates lots of test case data so that you can view it all in the app. Having an app flavour with debug functions is one idea but I'm not sure how scale-able this is as the number of test cases grows. Ideally I think having this generator code in the tests somehow would be helpful. The whole app is basically a front end for a SQL database so having code that just populates a database with a bunch of test data and graph configurations and then allows you to open that in the app might be the easiest way.

hhpmmd commented:
Having it as a local debug-branch in my repo is a good idea, didn't think of that. I reverted the commit in the PR branch.

So all that's left here is to check if we are happy with the hyper-parameters and then we're good to merge or am I missing something?

SamAmco commented:
The hyper-parameters looked like they were working pretty well to me so I merged :) You can always create a new PR if you want to improve it somehow but for now I think it's great! Thanks again @hhpmmd  I will try and releases this soon.
